src/main/java/org/datadog/jenkins/plugins/datadog/util/conversion/DatadogActionConverter.java

```diff
@@ -0,0 +1,115 @@
+     *                   no deserialization is performed.
+     */
+    @SafeVarargs
+    protected DatadogActionConverter(@Nonnull VersionedConverter<T>... converters) {
+        if (converters.length == 0) {
+            throw new IllegalArgumentException("At least one converter is needed");
+        }
+
+        readConverters = new HashMap<>();
+        for (VersionedConverter<T> converter : converters) {
+            VersionedConverter<T> existingConverter = readConverters.put(converter.getVersion(), converter);
+            if (existingConverter != null) {
+                throw new IllegalArgumentException(String.format("Two converters have the same version: %s (%d), %s (%d)",
+                        converter, converter.getVersion(), existingConverter, existingConverter.getVersion()));
+            }
+        }
+
+        writeConverter = readConverters.values()
+                .stream()
+                .max(Comparator.comparingInt(VersionedConverter::getVersion))
+                // this shouldn't be possible since readConverters must be non-empty
+                .orElseThrow(() -> new IllegalArgumentException("Cannot find converter with max version"));
+
+        // only direct children are supported, to keep logic simple
+        if (!getClass().getSuperclass().equals(DatadogActionConverter.class)) {
+            throw new IllegalArgumentException(getClass().getName() + " is not a direct descendant of " + DatadogActionConverter.class.getName());
+        }
+
+        ParameterizedType genericSuperclass = (ParameterizedType) getClass().getGenericSuperclass();
+        Type[] actualTypeArguments = genericSuperclass.getActualTypeArguments();
+        convertedType = (Class<?>) actualTypeArguments[0];
+    }
+
+    @Override
+    public boolean canConvert(Class type) {
+        return type.equals(convertedType);
+    }
+
+    @Override
+    public void marshal(Object source, HierarchicalStreamWriter writer, MarshallingContext context) {
+        writer.addAttribute(VERSION_ATTRIBUTE, String.valueOf(writeConverter.getVersion()));
+        writeConverter.marshal((T) source, writer, context);
+    }
+
+    @Nullable
+    @Override
+    public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
+        VersionedConverter<T> readConverter = getReadConverter(reader);
+        if (readConverter != null) {
+            return readConverter.unmarshal(reader, context);
+        } else {
+            // no matching converter found, data is too old or corrupted
+            return null;
+        }
+    }
+
+    @Nullable
+    private VersionedConverter<T> getReadConverter(HierarchicalStreamReader reader) {
+        try {
+            String versionString = reader.getAttribute(VERSION_ATTRIBUTE);
+            if (versionString == null) {
+                // no attribute, data was written by an old version of the plugin
+                return null;
```

Comment:
I guess we're checking `null` every time we get an action, right?

Reply:
Yes, these checks already exist: even before adding these changes it was possible to get a `null` simply because an action was never added to a run/node (e.g. if CI Visibility was enabled while the run was already in progress, or if whitelist of tracked jobs changed, etc)